packages/oats-runtime/src/make.ts

```diff
@@ -533,6 +445,145 @@ export function makeNullable(maker: any) {
   };
 }
 
+function isScalar(type: Type): boolean {
+  if (type.type === 'named') {
+    return isScalar(type.reference.definition);
+  }
+  return ['array', 'object', 'union', 'intersection', 'unknown'].indexOf(type.type) < 0;
+}
+
+function enumOptions(type: Type): number | null {
+  if (type.type === 'null') {
+    return 1;
+  }
+  if (type.type === 'named') {
+    return enumOptions(type.reference.definition);
+  }
+  if (
+    type.type === 'string' ||
+    type.type === 'boolean' ||
+    type.type === 'number' ||
+    type.type === 'integer'
+  ) {
+    return type.enum?.length || null;
+  }
+  return null;
+}
+
+function priority(v: ObjectType['properties'][string]) {
+  // check scalars first to avoid constructing trees unnecessarily
+  if (isScalar(v.value)) {
+    const enums = enumOptions(v.value);
+    if (enums !== null) {
+      // high chance this is a union type tag
+      if (enums === 1) {
+        return 4;
+      }
+      // enum matches less things than non enum
+      return 3;
+    }
+    if (v.value.type === 'string') {
+      // a pattern matches less things than a non pattern so lets try this first
+      if (v.value.pattern) {
+        return 3;
+      }
+    }
+    if (v.required) {
+      return 2;
+    }
+    return 1;
+  }
+  return 0;
+}
+
+function fromObjectReflection(type: ObjectType): Maker<any, any> {
+  const comparisonOrder = Object.keys(type.properties).sort((aKey, bKey) => {
+    const a = type.properties[aKey]!;
+    const b = type.properties[bKey]!;
+    return priority(b) - priority(a);
+  });
+  return makeObject(
+    Object.entries(type.properties).reduce(
+      (memo, [key, prop]) => ({
+        ...memo,
+        [key]: prop.required ? fromReflection(prop.value) : { optional: fromReflection(prop.value) }
+      }),
+      {}
+    ),
+    type.additionalProperties
+      ? type.additionalProperties === true
+        ? makeAny()
+        : fromReflection(type.additionalProperties)
+      : undefined,
+    comparisonOrder
+  );
+}
+
+function fromUnionReflection(types: Type[]): Maker<any, any> {
+  const { discriminator, undiscriminated } = discriminateUnion(types);
```

Comment:
this was not at all nice to write. Turns out that the previous discriminator code considered non singular enums as type tags which was a nice find.

Reply:
did this change the non-standard behaviour of our discriminator implementation? Meaning that before we accepted any discriminator value as long as it was unique per `oneOf` ref, but the OpenApi specification says that the value must contain the name of the schema

Reply:
> OpenApi specification says that the value must contain the name of the schema

thats the non-standard compliant bit I was kinda annoyed -- also somewhat annoyed at the standard itself. 

This pr deprecates the whole explicit discriminator support (but I need to add the runtime code for that still as otherwise this becomes non backwards compatible which sucks a bit)

Reply:
> also somewhat annoyed at the standard itself.

Indeed. The way it was defined made little sense to me 😅

> This pr deprecates the whole explicit discriminator support

Oh, I thought this was you adding it back due to the degraded performance issues you mentioned

Reply:
> but I need to add the runtime code for that still as otherwise this becomes non backwards compatible which sucks a bit

Reply:
added, but forgot to push 😬

Reply:
took me a while to realise what this new code actually did. clever 👌